dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLlvm.java

```diff
@@ -1,18 +1,1434 @@
+        // see https://llvm.org/docs/LangRef.html#cmpxchg-instruction
+        // TODO LLVM has no distinguished boolean type.
+        final Expression address = visitTypeValue(ctx.typeValue(0));
+        final Expression comparator = visitTypeValue(ctx.typeValue(1));
+        final Expression substitute = visitTypeValue(ctx.typeValue(2));
+        check(comparator.getType().equals(substitute.getType()), "Type mismatch for comparator and new in %s.", ctx);
+        final Register value = function.newRegister(comparator.getType());
+        final Register asExpected = function.newRegister(types.getBooleanType());
+        final boolean weak = ctx.weak != null;
+        final String mo = parseMemoryOrder(ctx.atomicOrdering(0));
+        block.events.add(newCompareExchange(value, asExpected, address, comparator, substitute, mo, weak));
+        final Register register = currentRegisterName == null ? null :
+                getOrNewCurrentRegister(types.getAggregateType(List.of(comparator.getType(), getIntegerType(1))));
+        if (register != null) {
+            final Expression cast = expressions.makeIntegerCast(asExpected, getIntegerType(1), false);
+            final Expression result = expressions.makeConstruct(List.of(value, cast));
+            block.events.add(newLocal(register, result));
+        }
+        return register;
+    }
+
+    @Override
+    public Expression visitAtomicRMWInst(AtomicRMWInstContext ctx) {
+        final Expression address = visitTypeValue(ctx.typeValue(0));
+        final Expression operand = visitTypeValue(ctx.typeValue(1));
+        final Type valueType = operand.getType();
+        final Register register = getOrNewCurrentRegister(valueType);
+        final String operator = ctx.atomicOp().getText();
+        final String mo = parseMemoryOrder(ctx.atomicOrdering());
+        final Event event;
+        if (operator.equals("xchg")) {
+            event = Llvm.newExchange(register, address, operand, mo);
+        } else {
+            final IOpBin op = switch (operator) {
+                case "add" -> IOpBin.PLUS;
+                case "sub" -> IOpBin.MINUS;
+                case "and" -> IOpBin.AND;
+                case "or" -> IOpBin.OR;
+                case "xor" -> IOpBin.XOR;
+                //TODO nand, min, umin, max, umax, uinc_wrap, udec_wrap, fadd, fsub, fmax, fmin
+                default -> throw new UnsupportedOperationException(String.format("Unknown atomic operand %s.", ctx.getText()));
+            };
+            event = Llvm.newRMW(register, address, operand, op, mo);
+        }
+        block.events.add(event);
+        return register;
+    }
+
+    // Conversions
+
+    @Override
+    public Expression visitTruncInst(TruncInstContext ctx) {
+        return conversionInstruction(ctx.typeValue(), ctx.type(), false);
+    }
+
+    @Override
+    public Expression visitZExtInst(ZExtInstContext ctx) {
+        return conversionInstruction(ctx.typeValue(), ctx.type(), false);
```

Comment:
We use a single conversion expression for both extending and truncating. The difference is only the target type (if it is smaller, then we truncate, otherwise we extend).